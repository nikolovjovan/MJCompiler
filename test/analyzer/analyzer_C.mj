// level C symbol detection:
// - abstract inner class
// - access to inner class field
// - inner class method call
program level_c_symbol_detection
    // const declarations
    const int a = 20, b = -1, c = +25;
    const char d = 'a';
    const bool e = true, f = false;

    // global variables
    int var_a, var_b[];
    char var_c, var_d[], var_e;
    bool var_f;

    abstract class abstract_class { // abstract inner class
        public int abstract_class_public_field;
        protected char abstract_class_protected_field;
        private bool abstract_class_private_field;
        {
            public abstract void public_method(int a, int b);
            protected abstract void protected_method(char a);
        }
    }

    class concrete_class extends abstract_class {
        // class variables
        public int class_var_a, class_var_b[];
        protected char class_var_c;
        private bool class_var_d;
        {
            // abstract method implementation
            public void public_method(int a, int b)
                int c;
            {
                abstract_class_public_field = 12; // superclass public field access
                abstract_class_protected_field = 'c'; // superclass protected field access
                class_var_a = 15; // this class public field access
                class_var_c = 'd'; // this class protected field access
                class_var_d = false; // this class private field access
            }
            // abstract method implementation
            protected void protected_method(char a) {}
            private void private_method(bool ind) {}
        }
    }

    class test_class {
        // class variables
        public int class_var_a, class_var_b[];
        protected char class_var_c;
        private bool class_var_d;
        {
            public void test_method() {}
        }
    }
{
    int calc(int a, int b, bool c) {
        if (c) return a + b;
        return a * b;
    }

    void test()
        // local variables
        int local_var_a, local_var_b[], i;
        char local_var_c;
        bool local_var_d, local_var_e[];
        test_class local_var_obj;
    {
        local_var_a = a; // int const usage
        local_var_b = new int[5]; // array instantiation
        local_var_c = d; // char const usage
        local_var_obj = new test_class;

        local_var_a = local_var_obj.class_var_a; // class field access
        local_var_b[3] = local_var_obj.class_var_b[5]; // class array field element access
        local_var_obj.test_method(); // class method call

        local_var_a = calc(local_var_a, local_var_a, local_var_d); // test method call
        local_var_b = new int[c - a]; // const expr
        for (i = 0; i < c - a; i++) {
            if (i % 2 == 0) local_var_c = 'e';
            else local_var_c = 'o';
            continue;
        }
    }

    void main()
        int a, b;
        char c;
        bool ind;
        abstract_class var;
    {
        a = 4;
        b = 15;
        c = 'b';
        ind = false;
        var = new concrete_class; // concrete class instantiation
        var.abstract_class_public_field = -5; // superclass public field access
        var.public_method(a, b); // public method call
    }
}