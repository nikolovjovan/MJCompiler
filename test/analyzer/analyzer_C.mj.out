Syntax analysis completed without errors.
Starting semantic analysis...
Found abstract class usage 'abstract_class' at line 26! Symbol node: Type 'abstract_class': type = 'abstract class', adr = 0, level = 0
Found class 'abstract_class' field usage 'abstract_class_public_field' at line 36! Symbol node: Field 'abstract_class_public_field': type = 'int', adr = 1, level = 1, parent = 'abstract_class', access modifier = public
Found class 'abstract_class' field usage 'abstract_class_protected_field' at line 37! Symbol node: Field 'abstract_class_protected_field': type = 'char', adr = 2, level = 1, parent = 'abstract_class', access modifier = protected
Found class 'concrete_class' field usage 'class_var_a' at line 38! Symbol node: Field 'class_var_a': type = 'int', adr = 4, level = 1, parent = 'concrete_class', access modifier = public
Found class 'concrete_class' field usage 'class_var_c' at line 39! Symbol node: Field 'class_var_c': type = 'char', adr = 6, level = 1, parent = 'concrete_class', access modifier = protected
Found class 'concrete_class' field usage 'class_var_d' at line 40! Symbol node: Field 'class_var_d': type = 'bool', adr = 7, level = 1, parent = 'concrete_class', access modifier = private
Found formal parameter usage 'c' at line 59! Symbol node: Formal parameter 'c': type = 'bool', adr = 2, level = 1
Found formal parameter usage 'a' at line 59! Symbol node: Formal parameter 'a': type = 'int', adr = 0, level = 1
Found formal parameter usage 'b' at line 59! Symbol node: Formal parameter 'b': type = 'int', adr = 1, level = 1
Found formal parameter usage 'a' at line 60! Symbol node: Formal parameter 'a': type = 'int', adr = 0, level = 1
Found formal parameter usage 'b' at line 60! Symbol node: Formal parameter 'b': type = 'int', adr = 1, level = 1
Found local variable usage 'local_var_a' at line 70! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found symbolic constant usage 'a' at line 70! Symbol node: Constant 'a': type = 'int', adr = 20, level = 0
Found local variable usage 'local_var_b' at line 71! Symbol node: Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
Found local variable usage 'local_var_c' at line 72! Symbol node: Variable 'local_var_c': type = 'char', adr = 3, level = 1
Found symbolic constant usage 'd' at line 72! Symbol node: Constant 'd': type = 'char', adr = 97, level = 0
Found local variable usage 'local_var_obj' at line 73! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found local variable usage 'local_var_a' at line 75! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_obj' at line 75! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found class 'test_class' field usage 'class_var_a' at line 75! Symbol node: Field 'class_var_a': type = 'int', adr = 1, level = 1, parent = 'test_class', access modifier = public
Found local variable usage 'local_var_b' at line 76! Symbol node: Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
Found array element access 'local_var_b[3]' at line 76! Symbol node: Array element 'local_var_b[3]': type = 'int', adr = -1, level = -1
Found local variable usage 'local_var_obj' at line 76! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found class 'test_class' field usage 'class_var_b' at line 76! Symbol node: Field 'class_var_b': type = 'array of int', adr = 2, level = 1, parent = 'test_class', access modifier = public
Found array element access 'local_var_obj.class_var_b[5]' at line 76! Symbol node: Array element 'local_var_obj.class_var_b[5]': type = 'int', adr = -1, level = -1
Found local variable usage 'local_var_obj' at line 77! Symbol node: Variable 'local_var_obj': type = 'class', adr = 6, level = 1
Found class 'test_class' field usage 'test_method' at line 77! Symbol node: Method 'test_method': type = 'void', adr = 0, level = 1, parent = 'test_class', access modifier = public
  Variable 'this': , adr = 0, level = 1
Found local variable usage 'local_var_a' at line 79! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found global method call 'calc' at line 79! Symbol node: Method 'calc': type = 'int', adr = 0, level = 3
  Formal parameter 'a': type = 'int', adr = 0, level = 1
  Formal parameter 'b': type = 'int', adr = 1, level = 1
  Formal parameter 'c': type = 'bool', adr = 2, level = 1
Found local variable usage 'local_var_a' at line 79! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_a' at line 79! Symbol node: Variable 'local_var_a': type = 'int', adr = 0, level = 1
Found local variable usage 'local_var_d' at line 79! Symbol node: Variable 'local_var_d': type = 'bool', adr = 4, level = 1
Found local variable usage 'local_var_b' at line 80! Symbol node: Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
Found symbolic constant usage 'c' at line 80! Symbol node: Constant 'c': type = 'int', adr = 25, level = 0
Found symbolic constant usage 'a' at line 80! Symbol node: Constant 'a': type = 'int', adr = 20, level = 0
Found local variable usage 'i' at line 81! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found local variable usage 'i' at line 81! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found symbolic constant usage 'c' at line 81! Symbol node: Constant 'c': type = 'int', adr = 25, level = 0
Found symbolic constant usage 'a' at line 81! Symbol node: Constant 'a': type = 'int', adr = 20, level = 0
Found local variable usage 'i' at line 81! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found for loop 'none' at line 81!
Found local variable usage 'i' at line 82! Symbol node: Variable 'i': type = 'int', adr = 2, level = 1
Found local variable usage 'local_var_c' at line 82! Symbol node: Variable 'local_var_c': type = 'char', adr = 3, level = 1
Found local variable usage 'local_var_c' at line 83! Symbol node: Variable 'local_var_c': type = 'char', adr = 3, level = 1
Found abstract class usage 'abstract_class' at line 92! Symbol node: Type 'abstract_class': type = 'abstract class', adr = 0, level = 0
Found local variable usage 'a' at line 94! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found local variable usage 'b' at line 95! Symbol node: Variable 'b': type = 'int', adr = 1, level = 1
Found local variable usage 'c' at line 96! Symbol node: Variable 'c': type = 'char', adr = 2, level = 1
Found local variable usage 'ind' at line 97! Symbol node: Variable 'ind': type = 'bool', adr = 3, level = 1
Found local variable usage 'var' at line 98! Symbol node: Variable 'var': type = 'class', adr = 4, level = 1
Found local variable usage 'var' at line 99! Symbol node: Variable 'var': type = 'class', adr = 4, level = 1
Found class 'abstract_class' field usage 'abstract_class_public_field' at line 99! Symbol node: Field 'abstract_class_public_field': type = 'int', adr = 1, level = 1, parent = 'abstract_class', access modifier = public
Found local variable usage 'var' at line 100! Symbol node: Variable 'var': type = 'class', adr = 4, level = 1
Found class 'abstract_class' field usage 'public_method' at line 100! Symbol node: Abstract method 'public_method': type = 'void', adr = 0, level = 3, parent = 'abstract_class', access modifier = public
  Variable 'this': , adr = 0, level = 1
  Formal parameter 'a': type = 'int', adr = 1, level = 1
  Formal parameter 'b': type = 'int', adr = 2, level = 1
Found local variable usage 'a' at line 100! Symbol node: Variable 'a': type = 'int', adr = 0, level = 1
Found local variable usage 'b' at line 100! Symbol node: Variable 'b': type = 'int', adr = 1, level = 1
Semantic analysis completed without errors.
======================================== SYMBOL TABLE DUMP =========================================
Type 'int': type = 'int', adr = -1, level = -1
Type 'char': type = 'char', adr = -1, level = -1
Type 'bool': type = 'bool', adr = -1, level = -1
Constant 'null': type = 'class', adr = 0, level = 0
Constant 'eol': type = 'char', adr = 10, level = 0
Method 'chr': type = 'char', adr = 0, level = 1
  Formal parameter 'i': type = 'int', adr = 0, level = 1
Method 'ord': type = 'int', adr = 0, level = 1
  Formal parameter 'ch': type = 'char', adr = 0, level = 1
Method 'len': type = 'int', adr = 0, level = 1
  Formal parameter 'arr': type = 'array of none', adr = 0, level = 1
Program 'level_c_symbol_detection': type = 'none', adr = 0, level = 1
  Constant 'a': type = 'int', adr = 20, level = 0
  Constant 'b': type = 'int', adr = -1, level = 0
  Constant 'c': type = 'int', adr = 25, level = 0
  Constant 'd': type = 'char', adr = 97, level = 0
  Constant 'e': type = 'bool', adr = 1, level = 0
  Constant 'f': type = 'bool', adr = 0, level = 0
  Variable 'var_a': type = 'int', adr = 0, level = 0
  Variable 'var_b': type = 'array of int', adr = 1, level = 0
  Variable 'var_c': type = 'char', adr = 2, level = 0
  Variable 'var_d': type = 'array of char', adr = 3, level = 0
  Variable 'var_e': type = 'char', adr = 4, level = 0
  Variable 'var_f': type = 'bool', adr = 5, level = 0
  Type 'abstract_class': type = 'abstract class', adr = 0, level = 0
  Type 'concrete_class': type = 'class', adr = 0, level = 0
  Type 'test_class': type = 'class', adr = 0, level = 0
  Method 'calc': type = 'int', adr = 0, level = 3
    Formal parameter 'a': type = 'int', adr = 0, level = 1
    Formal parameter 'b': type = 'int', adr = 1, level = 1
    Formal parameter 'c': type = 'bool', adr = 2, level = 1
  Method 'test': type = 'void', adr = 0, level = 0
    Variable 'local_var_a': type = 'int', adr = 0, level = 1
    Variable 'local_var_b': type = 'array of int', adr = 1, level = 1
    Variable 'i': type = 'int', adr = 2, level = 1
    Variable 'local_var_c': type = 'char', adr = 3, level = 1
    Variable 'local_var_d': type = 'bool', adr = 4, level = 1
    Variable 'local_var_e': type = 'array of bool', adr = 5, level = 1
    Variable 'local_var_obj': type = 'class', adr = 6, level = 1
  Method 'main': type = 'void', adr = 0, level = 0
    Variable 'a': type = 'int', adr = 0, level = 1
    Variable 'b': type = 'int', adr = 1, level = 1
    Variable 'c': type = 'char', adr = 2, level = 1
    Variable 'ind': type = 'bool', adr = 3, level = 1
    Variable 'var': type = 'class', adr = 4, level = 1

